lib/grp.gi

```diff
@@ -4291,86 +4294,121 @@ end );
 InstallMethod( GroupWithGenerators,
     "generic method for collection",
     [ IsCollection ],
-    function( gens )
-    local G,fam,typ;
-
-    fam:=FamilyObj(gens);
-    if IsFinite(gens) then
-      if not IsBound(fam!.defaultFinitelyGeneratedGroupType) then
-        fam!.defaultFinitelyGeneratedGroupType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses
-                      and IsFinitelyGeneratedGroup);
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupType;
-    else
-      if not IsBound(fam!.defaultGroupType) then
-        fam!.defaultGroupType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses);
+function( gens )
+local G,fam,typ,id,triv;
+
+  fam:=FamilyObj(gens);
+
+  triv:=fail; # might not find out
+
+  typ:=IsGroup and IsAttributeStoringRep 
+    and HasIsEmpty and HasGeneratorsOfMagmaWithInverses;
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    if Length(gens)<=1 then
+      typ:=typ and HasIsCyclic and IsCyclic 
+          and HasIsCommutative and IsCommutative;
+    fi;
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      id:=One(gens[1]);
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=false;
+      else
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=true;
+        # basic consequence
+        typ:=typ and HasIsPerfectGroup and IsPerfectGroup;
       fi;
-      typ:=fam!.defaultGroupType;
     fi;
 
-    G:=rec();
-    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
+  fi;
+  typ:=NewType(fam,typ);
 
-    return G;
-    end );
+  G:=rec();
+  if triv=fail then
+    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens),
+                            IsEmpty,false);
+  else
+    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens),
+                            IsEmpty,false,IsTrivial,triv,
+                            IsNonTrivial,not triv);
```

Comment:
Hmm, why set `IsTrivial` resp. `IsNonTrivial` this way, instead of just anding them into `typ` like you do with e.g. `IsCyclic`? Does this have some functional difference?

If not, I'd just set this all in `typ`, then you can get rid of `triv` and simplify the code here and elsewhere accordingly.

Reply:
I'd love to do. But how do I set  a flag to `false' as part of a type?

IsTrivial and IsNonTrivial are twins that are set the opposite way and currently immediate methods are used to set either to false if the other is set.

Reply:
Thanks! Will do so.